libyul/optimiser/OptimiserStep.h

```diff
@@ -29,12 +29,12 @@ namespace solidity::yul
 
 struct Dialect;
 struct Block;
-class NameDispenser;
+class YulNameRepository;
 
 struct OptimiserStepContext
 {
 	Dialect const& dialect;
-	NameDispenser& dispenser;
+	YulNameRepository& nameRepository;
```

Comment:
We should not in particular carry the name repository around here.
The main advantage of having introduced an outer ``AST`` struct around the blocks is that this ``AST`` struct can contain arbitrary auxiliary information about the AST, including ownership of AST nodes in the future, etc.
So ideally, it would be a reference to the ``AST`` struct itself here in the context - otherwise we'll only need to refactor this again later on.

Reply:
Or I'd at least think about that. I.e. we're moving towards having proper context state here already - so it may be worthwhile to think ahead towards e.g. having a memory management data structure for AST nodes in here as well. The one problem with actually putting an ``AST`` reference here is that potentially one may want to distinguish between steps that modify an `AST` and steps that consume one, recreating a new one. (It would also be an interesting question whether all steps should only have one of those two forms anyways.)